opt/test/test_loom.py

```diff
@@ -0,0 +1,145 @@
+    with lp.connect(LOOM_PATH, mode="r", validate=False) as ds:
+        test_loom = Loom(LOOM_PATH, LOOM_PATH, ds, LOOM_FILE_HANDLER)
+        assert test_loom.get_global_attribute_by_name("Genome") == "Nomen dubium"
+
+
+def test_get_cell_ids(loom_file):
+    matrix, row_attrs, col_attrs, attrs = loom_file
+    lp.create(filename=str(LOOM_PATH), layers=matrix, row_attrs=row_attrs, col_attrs=col_attrs, file_attrs=attrs)
+    with lp.connect(LOOM_PATH, mode="r", validate=False) as ds:
+        test_loom = Loom(LOOM_PATH, LOOM_PATH, ds, LOOM_FILE_HANDLER)
+        num_cells = ds.shape[1]
+        assert (test_loom.get_cell_ids() == np.array([f"Cell_{n}" for n in range(1, num_cells + 1)])).all()
+
+
+def test_get_gene_names(loom_file):
+    matrix, row_attrs, col_attrs, attrs = loom_file
+    lp.create(filename=str(LOOM_PATH), layers=matrix, row_attrs=row_attrs, col_attrs=col_attrs, file_attrs=attrs)
+    with lp.connect(LOOM_PATH, mode="r", validate=False) as ds:
+        test_loom = Loom(LOOM_PATH, LOOM_PATH, ds, LOOM_FILE_HANDLER)
+        assert test_loom.get_gene_names() == {}
+
+
+def test_get_meta_data_cluster_by_clustering_id_and_cluster_id(loom_file):
+    matrix, row_attrs, col_attrs, attrs = loom_file
+    lp.create(filename=str(LOOM_PATH), layers=matrix, row_attrs=row_attrs, col_attrs=col_attrs, file_attrs=attrs)
+    with lp.connect(LOOM_PATH, mode="r", validate=False) as ds:
+        test_loom = Loom(LOOM_PATH, LOOM_PATH, ds, LOOM_FILE_HANDLER)
+        assert (
+            test_loom.get_meta_data_cluster_by_clustering_id_and_cluster_id(0, 0)["description"]
+            == "Unannotated Cluster 1"
+        )
+
+
+def test_infer_species(loom_file):
+    matrix, row_attrs, col_attrs, attrs = loom_file
+    lp.create(filename=str(LOOM_PATH), layers=matrix, row_attrs=row_attrs, col_attrs=col_attrs, file_attrs=attrs)
+    with lp.connect(LOOM_PATH, mode="r", validate=False) as ds:
+        test_loom = Loom(LOOM_PATH, LOOM_PATH, ds, LOOM_FILE_HANDLER)
+        assert test_loom.infer_species() == ("Unknown", {})
+
+
+def test_has_motif_and_track_regulons(loom_file):
+    matrix, row_attrs, col_attrs, attrs = loom_file
+    lp.create(filename=str(LOOM_PATH), layers=matrix, row_attrs=row_attrs, col_attrs=col_attrs, file_attrs=attrs)
+    with lp.connect(LOOM_PATH, mode="r", validate=False) as ds:
+        test_loom = Loom(LOOM_PATH, LOOM_PATH, ds, LOOM_FILE_HANDLER)
+        assert test_loom.has_motif_and_track_regulons() == True
+
+
+def test_get_coordinates(loom_file):
+    matrix, row_attrs, col_attrs, attrs = loom_file
+
+    # POC for custom data in the loom file.
+    num_cells = 100
+    _X = np.concatenate([rg.normal(n, 0.1, int(num_cells / 4)) for n in range(-2, 2)])
+    _Y = rg.normal(0, 0.1, num_cells)
+    main_embedding = pd.DataFrame(columns=["_X", "_Y"])
+    main_embedding["_X"] = _X
+    main_embedding["_Y"] = _Y
+    col_attrs["Embedding"] = Loom.dfToNamedMatrix(main_embedding)
```

Comment:
Shouldn't this be imported from `opt/scopeserver/gen_test_loom.py` ? Seems to be like redundant code

Reply:
This actually generates different results but is also a proof of concept for generating data in the test to use in the loom file.